chains/chains.go

```diff
@@ -123,46 +124,82 @@ func Predict(ctx context.Context, c Chain, inputValues map[string]any, options .
 
 const _defaultApplyMaxNumberWorkers = 5
 
+type applyInputJob struct {
+	input map[string]any
+	i     int
+}
+
+type applyResult struct {
+	result map[string]any
+	err    error
+	i      int
+}
+
 // Apply executes the chain for each of the inputs asynchronously.
 func Apply(ctx context.Context, c Chain, inputValues []map[string]any, maxWorkers int, options ...ChainCallOption) ([]map[string]any, error) { // nolint:lll
 	if maxWorkers <= 0 {
 		maxWorkers = _defaultApplyMaxNumberWorkers
 	}
 
-	inputJobs := make(chan map[string]any, len(inputValues))
-	resultsChan := make(chan struct {
-		result map[string]any
-		err    error
-	}, len(inputValues))
-	defer close(inputJobs)
-	defer close(resultsChan)
+	inputJobs := make(chan applyInputJob, len(inputValues))
+	resultsChan := make(chan applyResult, len(inputValues))
+
+	var wg sync.WaitGroup
+	wg.Add(maxWorkers)
 
 	for w := 0; w < maxWorkers; w++ {
 		go func() {
-			for input := range inputJobs {
-				res, err := Call(ctx, c, input, options...)
-				resultsChan <- struct {
-					result map[string]any
-					err    error
-				}{
-					result: res,
-					err:    err,
+			defer wg.Done()
+			for {
+				select {
+				case <-ctx.Done():
+					return
```

Comment:
We can still write to `resultsChan` here with the error returned by `ctx.Err` i.e.
```go
resultsChan <- applyResult{
    err: ctx.Err()
    // can add a sentinel value for `i` e.g. -1 if needed, but don't think it's that important atm
}
```
That enables the client to know if the `Apply` function failed due to a timeout/cancellation.

Reply:
Ignore - I've just seen `getApplyResults` 😄

Reply:
There’s also cancel reason in newer go releases.